File: packages/engine/src/cpus/6502/instructions/brk.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('brk', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.sp = 0xff
  cpu.interruptDisabled = false
  mmu.data = [0x0] // BRK
  mmu.data[0xfffe] = 0x04
  mmu.data[0xffff] = 0xac
  const g = cpu.run()
  for (let i = 0; i < 7; i++) g.next()
  expect(cpu.pc).toBe(0x4ac)
  expect(mmu.data[0x1ff]).toBe(0)
  expect(mmu.data[0x1fe]).toBe(2)
  expect(mmu.data[0x1fd]).toBe((cpu.status | 0x10) & ~0x4)
})

test('brk hijacked', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.sp = 0xff
  cpu.interruptDisabled = false
  mmu.data = [0x0] // BRK
  mmu.data[0xfffa] = 0x04
  mmu.data[0xfffb] = 0xac
  const g = cpu.run()
  for (let i = 0; i < 7; i++) {
    if (i === 1) cpu.nmiRequested = true
    g.next()
  }
  expect(cpu.pc).toBe(0x4ac)
  expect(mmu.data[0x1ff]).toBe(0)
  expect(mmu.data[0x1fe]).toBe(2)
  expect(mmu.data[0x1fd]).toBe((cpu.status | 0x10) & ~0x4)
})

File: packages/engine/src/cores/famibob/cpu/instructions/brk.test.zig
const testing = @import("../testing/setup.zig");

test "BRK" {
    var cpu = testing.createCPU();
    cpu.sp = 0xff;
    cpu.i = false;
    cpu.write(0, 0);
    cpu.write(0xfffe, 4);
    cpu.write(0xffff, 0xac);
    for (0..7) |_| {
        cpu.process();
    }
    try testing.expectEqual(0x4ac, cpu.pc);
    try testing.expectEqual(0, cpu.read(0x1ff));
    try testing.expectEqual(2, cpu.read(0x1fe));
    try testing.expectEqual((cpu.getP() | 0x10) & ~@as(u8, 0x4), cpu.read(0x1fd));
    try testing.expectEqual(true, cpu.i);
}

test "BRK hijacked" {
    var cpu = testing.createCPU();
    cpu.sp = 0xff;
    cpu.i = false;
    cpu.write(0, 0);
    cpu.write(0xfffa, 4);
    cpu.write(0xfffb, 0xac);
    for (0..7) |i| {
        if (i == 1) cpu.nmi_requested = true;
        cpu.process();
    }
    try testing.expectEqual(0x4ac, cpu.pc);
    try testing.expectEqual(0, cpu.read(0x1ff));
    try testing.expectEqual(2, cpu.read(0x1fe));
    try testing.expectEqual((cpu.getP() | 0x10) & ~@as(u8, 0x4), cpu.read(0x1fd));
    try testing.expectEqual(true, cpu.i);
}
